### PR TITLE
Ignore indexer properties

### PR DESCRIPTION
**Pull request by MarcoRossignoli**
closes https://github.com/dotnet/corefx/issues/38131

cc: @ahsonkhan @steveharter

### CODE REVIEW

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs [hunk 1]
```diff
@@ -97,6 +97,12 @@ internal JsonClassInfo(Type type, JsonSerializerOptions options)
 
                     foreach (PropertyInfo propertyInfo in type.GetProperties(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic))
                     {
+                        // We ignore public indexer
```
Comment (stephentoub) [inline src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs]: Nit: all indexers are ignored... it's a little strange to me to just call out "public" here.  I'd suggest just "Ignore indexers".

File: src/System.Text.Json/tests/Serialization/Object.ReadTests.cs [hunk 2]
```diff
@@ -230,5 +230,12 @@ public static void ReadConversionFails()
 
             Assert.True(exceptionThrown);
         }
+
+        [Fact]
+        public static void ReadObject_PublicIndexer()
+        {
+            var indexer = JsonSerializer.Parse<Indexer>(@"{""NonIndexerProp"":""Value""}");
```
Comment (stephentoub) [inline src/System.Text.Json/tests/Serialization/Object.ReadTests.cs]: Nit: var => Indexer

File: src/System.Text.Json/tests/Serialization/Object.WriteTests.cs [hunk 3]
```diff
@@ -53,5 +53,27 @@ public class ObjectObject
...
+        [Fact]
+        public static void WriteObject_PublicIndexer()
+        {
+            var indexer = new Indexer();
+            indexer[42] = default;
+            indexer.NonIndexerProp = "Value";
+            Assert.Equal(@"{""NonIndexerProp"":""Value""}", JsonSerializer.ToString(indexer));
+        }
+
+        class Indexer
```
Comment (stephentoub) [inline src/System.Text.Json/tests/Serialization/Object.WriteTests.cs]: Nit: private

File: src/System.Text.Json/tests/Serialization/Object.WriteTests.cs [hunk 4]
```diff
@@ -53,5 +53,27 @@ public class ObjectObject
         {
             public object Object { get; set; }
         }
+
+        [Fact]
+        public static void WriteObject_PublicIndexer()
+        {
+            var indexer = new Indexer();
+            indexer[42] = default;
```
Comment (stephentoub) [inline src/System.Text.Json/tests/Serialization/Object.WriteTests.cs]: Consider setting it to something other than 0.

File: src/System.Text.Json/tests/Serialization/Object.ReadTests.cs (hunk 2 above)
Comment (MarcoRossignoli) [reply src/System.Text.Json/tests/Serialization/Object.ReadTests.cs]: I thought that parse generics arg was enough clear, will fix

File: src/System.Text.Json/tests/Serialization/Object.WriteTests.cs (hunk 4 above)
Comment (MarcoRossignoli) [reply src/System.Text.Json/tests/Serialization/Object.WriteTests.cs]: right.

File: src/System.Text.Json/tests/Serialization/Object.WriteTests.cs (hunk 3 above)
Comment (MarcoRossignoli) [reply src/System.Text.Json/tests/Serialization/Object.WriteTests.cs]: yep

**Comment by stephentoub** [comment]
Thanks.

**Comment by MarcoRossignoli** [comment]
unrelated fail https://mc.dot.net/#/user/dotnet-bot/pr~2Fdotnet~2Fcorefx~2Frefs~2Fpull~2F38133~2Fmerge/test~2Ffunctional~2Fcli~2Finnerloop~2F/20190603.3

File: src/System.Text.Json/tests/Serialization/Object.ReadTests.cs [hunk 5]
```diff
@@ -230,5 +230,12 @@ public static void ReadConversionFails()
 
             Assert.True(exceptionThrown);
         }
+
+        [Fact]
+        public static void ReadObject_PublicIndexer()
+        {
+            Indexer indexer = JsonSerializer.Parse<Indexer>(@"{""NonIndexerProp"":""Value""}");
+            Assert.Equal("Value", indexer.NonIndexerProp);
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/Object.ReadTests.cs]: nit: I would set this `_index` to some non-zero sentinel like -1 (within the `Indexer` class) and then verify that we are not clearing it on Parse.

So, add:
```C#
Assert.Equal(-1, indexer[0]);
```

File: src/System.Text.Json/tests/Serialization/Object.WriteTests.cs [hunk 6]
```diff
@@ -53,5 +53,27 @@ public class ObjectObject
...
             public object Object { get; set; }
         }
+
+        [Fact]
+        public static void WriteObject_PublicIndexer()
+        {
+            var indexer = new Indexer();
+            indexer[42] = 42;
+            indexer.NonIndexerProp = "Value";
+            Assert.Equal(@"{""NonIndexerProp"":""Value""}", JsonSerializer.ToString(indexer));
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/Object.WriteTests.cs]: nit: 

We are not consistent here across the test set, but we should generally try to test all the public overloads (which includes `ToUtf8Bytes`, and the `WriteAsync` API). The same applies on the read side. Though, I wouldn't single this PR/change out for that. This is something we need to do all up, so just pointing it out on new tests we are adding.

cc @layomia, @buyaa-n, @JeremyKuhne

File: src/System.Text.Json/tests/Serialization/Object.ReadTests.cs (hunk 5 above)
Comment (MarcoRossignoli) [reply src/System.Text.Json/tests/Serialization/Object.ReadTests.cs]: done

**Comment by MarcoRossignoli** [comment]
guys let me know if there is other work to do on this.

**Pull request merged**